### ISSUE TITLE
Feature Request: Allow Hilt to consume Dagger `@Component`s

### ISSUE THREAD

**Issue by gmazzo** [google/dagger#2994]
# Background
In my company, we expose our internal Android libraries as Dagger `@Components`. The rationally behind this pattern is that:
- we want to avoid conflicting binding (by exposing `@Modules` instead) 
- we want a clear declaration of each library's dependencies

In Hilt docs, [Custom Components](https://dagger.dev/hilt/custom-components) section, you state:
> Consider if a non-Hilt (regular Dagger) component is sufficient

Still that use case seems not to be supported (yet) because:
1) I couldn't find an API to property import a Dagger component (without re-writing all providers again). An equivalent to https://dagger.dev/api/2.22/dagger/Component.html#dependencies--
2) If by using Hilt compiler, you declare a `@Module` without an `@InstallIn` annotation (which will be the case for Dagger components), ~it fails~ (you can actually use `@DisableInstallInCheck` for this)

# Use case
Allow Hilt components to somehow consume (inherit its exposed providers) Dagger `@Component`s. If this is supported already, I couldn't find any documentation or example about it.

For instance, one way we may support this is by providing a component in a Hilt module:
```kotlin
@Module
@InstallsOn(SingletonComponent::class)
object AppModule {

    // by providing an instance of `LoggerComponent`, Hilt will be smart enough to generate a provider `Logger` from it
    @Provides
    fun provideLoggerComponent(application: Application): LoggerComponent =
        DaggerLoggerComponent.create(application)

}

class MainActivity : Activity {

    @Inject
    lateinit var logger: Logger

    fun onCreate(savedInstanceState: Bundle) {
        super.onCreate(savedInstanceState)
        
        logger.log("onCreate")
    }

}
```

And from our library module:
```kotlin
@Component
interface LoggerComponent {

    val logger: Logger
    
    @Component.Factory
    interface Factory {
    
        fun create(@BindsInstance context: Context)
    
    }

}

interface Logger {

    fun log(log: String)
    
}
```

**Comment by trevjonez** [comment]
Feels fairly similar to what I was thinking of as a POC/workaround in this request.
https://github.com/google/dagger/issues/2014
Where did all the time go :disappointed:

**Comment by Chang-Eric** [comment]
I think you could easily do this yourself by just writing an `@Provides` for `Logger` that takes your component as the dependency and returns the logger from it?

If you wanted it to be automatic for every type exposed on the interface, it shouldn't be too hard to write your own processor and code generator that outputs a Hilt module that does that after examining your component. It could be a nice extension library building off of Hilt.

**Comment by gmazzo** [comment]
> I think you could easily do this yourself by just writing an `@Provides` for `Logger` that takes your component as the dependency and returns the logger from it?

That's exactly our workaround, yes. 

> If you wanted it to be automatic for every type exposed on the interface, it shouldn't be too hard to write your own processor and code generator that outputs a Hilt module that does that after examining your component. It could be a nice extension library building off of Hilt.

We may build a tooling like that, yes. I just wanted to check if this was supported somehow already.

**Comment by Chang-Eric** [comment]
Ah yea, understood. We don't have anything like that today in Hilt, so I think building something might be your best bet. To be honest, I've built this before for a project I had to migrate, but there ended up being enough special case code and assumptions that it probably is better to just have something be use-case specific, especially since it doesn't need any special hooks into Hilt.

**Comment by gmazzo** [comment]
Then I'm going to close the issue. Feel free to re-open it the scenario changes in the future. Thanks!

**Comment by gmazzo** [comment]
By the way @Chang-Eric, we are drafting a new annotation + compiler that will take 1 o more `@Component interface`s from Dagger, and create a Hilt `@Module @InstallIn` with `@Providers` methods from it.   

Would you be interested in a contribution? We probably will be building it anyway internally.

Imagine something like:
```java
public @interface ImportDaggerComponent {
  Class<?>[] value();
}
```
Used at:
```java
@Module
@InstallIn(SingletonComponent.class)
@ImportDaggerComponent(LoggerComponent.class)
interface AppModule
```

The compiler output will be something like this:
```java
@Module
@DaggerGenerated
@InstallIn(SingletonComponent.class) // copied from `AppModule`
abstract class AppModule_DaggerComponentImports {

    @Provides
    static Logger provideLogger(LoggerComponent component) {
        return component.logger();
    }

}
```

Those `@Provide`s will introduce a new dependency requirement on a `LoggerComponent`, that the final user will be responsible to provide. For instance, in this hypothetical case by adding:
```java
@Module
@InstallIn(SingletonComponent.class)
@ImportDaggerComponent(LoggerComponent.class)
abstract class AppModule {
    
    @Provides
    static LoggerComponent provideLoggerComponent(Application application) {
        return DaggerLoggerComponent.create(application);
    }

}
```
All `@Provides` will be generated reading all no-args methods from the target `@Component interface`s, then it will be a generic solution. Unless I misted something 🤔

**Comment by danysantiago** [comment]
Hi - If you open source it I will gladly take a look.

Two things to remember:
* Annotated your annotation API (`@ImportDaggerComponent`) with `@GeneratesRootInput` and remember to apply the `hilt-compiler` to the runtime project where this API is.
* Annotate the generated `@Module` with `@OriginatingElement`

Both of these are explained here: https://dagger.dev/hilt/creating-extensions

Also, plz don't use `@DaggerGenerated`, that is internal not meant to be use by external code gens.

**Comment by Chang-Eric** [comment]
> Would you be interested in a contribution? We probably will be building it anyway internally.

Just to clarify, this is asking about contributing into Hilt? If so, I think probably not just because I'm not sure our team has the bandwidth to maintain it. There's a general problem here which is I'm kind of not sure yet how to handle these sort of side libraries that aren't core in Hilt but use or work together with Hilt.

Part of the maintenance I'm thinking about is we ran into issues internally where we had to start expanding the API. For example, we soon had to start adding a filter mechanism. This is because the components being pulled in were also used with component dependencies with other components, and so sometimes there were conflicts due needing to expose bindings for the other component dependency graph. This leads to conflicting bindings in the Hilt component even if the true source of the binding is the same. This happened a fair amount during migration. There's also a question of what to do if you want something with optional or multibindings which I think would need some extra declarations and stuff.

This is not to say these are blockers or are things that shouldn't be done. But mostly just that our team doesn't have the time to responsibly own such a feature right now. With that said, if you're running into something or want an opinion on a design choice, I'd be happy to look over what you've built and give thoughts.
